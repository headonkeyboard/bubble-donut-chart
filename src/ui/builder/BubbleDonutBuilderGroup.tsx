import React, { ChangeEvent, FunctionComponent, useCallback, useRef } from "react";

type BubbleDonutBuilderGroupProps = {
  index: number;
  defaultSize: number;
  color: string;
  onChange: (sectionIndex: number, size: number) => void;
};

const BubbleDonutBuilderGroup: FunctionComponent<BubbleDonutBuilderGroupProps> = ({
  index,
  defaultSize,
  color,
  onChange,
}) => {
  const groupSizeRef = useRef<number>(defaultSize);

  const handleSliderChange = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      groupSizeRef.current = parseInt(e.target.value, 10);

      onChange(index, groupSizeRef.current);
    },
    [index, onChange]
  );

  return (
    <div className="space-y-2">
      <div className="flex space-x-4 items-center">
        <label htmlFor={`group_${index}`} className="flex items-center space-x-2">
            <span
                className="block flex-shrink-0 w-3 h-3 rounded-full"
                style={{
                    backgroundColor: color,
                }}
            />
            <span>Group #{index + 1}</span>
        </label>
        <input
          className="flex-1"
          id={`group_${index}`}
          name="size"
          onChange={handleSliderChange}
          type="range"
          value={defaultSize}
          min="1"
          max="100"
        />
      </div>
    </div>
  );
};

export default BubbleDonutBuilderGroup;
