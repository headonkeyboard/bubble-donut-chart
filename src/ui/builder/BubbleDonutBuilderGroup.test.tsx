import React from "react";
import renderer from "react-test-renderer";
import {fireEvent, render} from "@testing-library/react";
import BubbleDonutBuilderGroup from './BubbleDonutBuilderGroup';

describe('BubbleDonutBuilderGroup', () => {
    test('BubbleDonutBuilderGroup should display a slider input with a label', () => {
        const component = renderer.create(
            <BubbleDonutBuilderGroup index={1} defaultSize={100} color="#EEEEEE" onChange={() => {}} />
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('BubbleDonutBuilderGroup onChange function should receive new slider value after each slider change', () => {
        const GROUP_INDEX = 1;
        const SLIDER_INITIAL_VALUE = 100;
        const TRIGGER_SLIDER_NEW_VALUE = 20;

        const onChangeCallback = jest.fn();

        const {container} = render(
            <BubbleDonutBuilderGroup index={GROUP_INDEX} defaultSize={SLIDER_INITIAL_VALUE} color="#EEEEEE" onChange={onChangeCallback} />
        );

        const slider = container.querySelector('input[type=range]');
        fireEvent.change(slider, {
            target: { value: 20 }
        });

        expect(onChangeCallback).toHaveBeenCalledWith(GROUP_INDEX, TRIGGER_SLIDER_NEW_VALUE);
    });
});
