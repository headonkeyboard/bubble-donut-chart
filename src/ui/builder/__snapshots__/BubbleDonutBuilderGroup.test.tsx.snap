// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BubbleDonutBuilderGroup BubbleDonutBuilderGroup should display a slider input with a label 1`] = `
<div
  className="space-y-2"
>
  <div
    className="flex space-x-4 items-center"
  >
    <label
      className="flex items-center space-x-2"
      htmlFor="group_1"
    >
      <span
        className="block flex-shrink-0 w-3 h-3 rounded-full"
        style={
          Object {
            "backgroundColor": "#EEEEEE",
          }
        }
      />
      <span>
        Group #
        2
      </span>
    </label>
    <input
      className="flex-1"
      id="group_1"
      max="100"
      min="1"
      name="group_size_1"
      onChange={[Function]}
      type="range"
      value={100}
    />
  </div>
</div>
`;
